File: packages/engine/src/liveEffects/quality.ts
/**
 * Render quality tiers for live effects.
 *
 * Three tiers with a strict contract:
 *   interactive — reduced internal resolution / sample counts. Only used while
 *                 the user is actively dragging a control; never persisted.
 *   normal      — full resolution, standard samples. Live-canvas default.
 *   export      — full resolution, maximum samples. Forced on all export paths.
 *
 * The serialized `quality` parameter on an effect is one of
 * 'auto' | 'interactive' | 'normal' | 'export'. 'auto' resolves to the
 * caller-provided tier, so documents that never set an explicit tier still
 * export at full quality and preview at normal quality. A user-selected tier
 * is honoured in preview but never degrades export (export callers always
 * pass 'export').
 */

export type EffectQuality = 'interactive' | 'normal' | 'export';

export type EffectQualityParam = 'auto' | 'interactive' | 'normal' | 'export';

/** Resolve a serialized quality param against the caller's tier. */
export function resolveEffectQuality(
  param: EffectQualityParam | undefined,
  caller: EffectQuality,
): EffectQuality {
  if (param && param !== 'auto') return param;
  return caller;
}

/** Internal resolution factor for an effective tier (1 = full res). */
export function qualityResolutionFactor(quality: EffectQuality): number {
  switch (quality) {
    case 'interactive':
      return 0.5;
    case 'normal':
    case 'export':
      return 1;
  }
}

/** Sample-count multiplier for a tier (relative to normal = 1). */
export function qualitySampleFactor(quality: EffectQuality): number {
  switch (quality) {
    case 'interactive':
      return 0.5;
    case 'normal':
      return 1;
    case 'export':
      return 2;
  }
}

/** Box-average downsample by an integer factor (deterministic, no canvas). */
export function downsampleBox(
  src: Uint8ClampedArray,
  w: number,
  h: number,
  factor: number,
): { data: Uint8ClampedArray<ArrayBuffer>; width: number; height: number } {
  const f = Math.max(1, Math.floor(factor));
  const dw = Math.max(1, Math.floor(w / f));
  const dh = Math.max(1, Math.floor(h / f));
  const out = new Uint8ClampedArray(new ArrayBuffer(dw * dh * 4));
  for (let y = 0; y < dh; y += 1) {
    for (let x = 0; x < dw; x += 1) {
      let r = 0;
      let g = 0;
      let b = 0;
      let a = 0;
      let n = 0;
      const y0 = y * f;
      const x0 = x * f;
      for (let sy = y0; sy < Math.min(y0 + f, h); sy += 1) {
        for (let sx = x0; sx < Math.min(x0 + f, w); sx += 1) {
          const o = (sy * w + sx) * 4;
          r += src[o]!;
          g += src[o + 1]!;
          b += src[o + 2]!;
          a += src[o + 3]!;
          n += 1;
        }
      }
      const o = (y * dw + x) * 4;
      out[o] = Math.round(r / n);
      out[o + 1] = Math.round(g / n);
      out[o + 2] = Math.round(b / n);
      out[o + 3] = Math.round(a / n);
    }
  }
  return { data: out, width: dw, height: dh };
}

/** Bilinear-upsample a buffer into a destination sized w×h (deterministic). */
export function upsampleBilinear(
  src: Uint8ClampedArray,
  sw: number,
  sh: number,
  dst: Uint8ClampedArray,
  w: number,
  h: number,
): void {
  const sxScale = sw / w;
  const syScale = sh / h;
  for (let y = 0; y < h; y += 1) {
    const sy = Math.min(sh - 1, Math.max(0, (y + 0.5) * syScale - 0.5));
    const y0 = Math.floor(sy);
    const y1 = Math.min(sh - 1, y0 + 1);
    const fy = sy - y0;
    for (let x = 0; x < w; x += 1) {
      const sx = Math.min(sw - 1, Math.max(0, (x + 0.5) * sxScale - 0.5));
      const x0 = Math.floor(sx);
      const x1 = Math.min(sw - 1, x0 + 1);
      const fx = sx - x0;
      const o00 = (y0 * sw + x0) * 4;
      const o10 = (y0 * sw + x1) * 4;
      const o01 = (y1 * sw + x0) * 4;
      const o11 = (y1 * sw + x1) * 4;
      const do_ = (y * w + x) * 4;
      for (let c = 0; c < 4; c += 1) {
        const top = src[o00 + c]! + (src[o10 + c]! - src[o00 + c]!) * fx;
        const bot = src[o01 + c]! + (src[o11 + c]! - src[o01 + c]!) * fx;
        dst[do_ + c] = Math.round(top + (bot - top) * fy);
      }
    }
  }
}

File: crates/varve-effects/src/quality.rs
//! Render quality tiers and resample helpers — faithful port of
//! `packages/engine/src/liveEffects/quality.ts`.

use crate::js_round;

/// Resolution factor for an effective tier (1 = full res).
pub fn quality_resolution_factor(quality: crate::EffectQuality) -> f64 {
    match quality {
        crate::EffectQuality::Interactive => 0.5,
        crate::EffectQuality::Normal | crate::EffectQuality::Export => 1.0,
    }
}

/// Sample-count multiplier for a tier (relative to normal = 1).
pub fn quality_sample_factor(quality: crate::EffectQuality) -> f64 {
    match quality {
        crate::EffectQuality::Interactive => 0.5,
        crate::EffectQuality::Normal => 1.0,
        crate::EffectQuality::Export => 2.0,
    }
}

/// Box-average downsample by an integer factor (deterministic, no canvas) —
/// port of `downsampleBox`.
pub fn downsample_box(src: &[u8], w: u32, h: u32, factor: u32) -> (Vec<u8>, u32, u32) {
    let f = factor.max(1);
    let dw = (w / f).max(1);
    let dh = (h / f).max(1);
    let mut out = vec![0u8; (dw * dh * 4) as usize];
    for y in 0..dh {
        for x in 0..dw {
            let mut r = 0u64;
            let mut g = 0u64;
            let mut b = 0u64;
            let mut a = 0u64;
            let mut n = 0u64;
            let y0 = y * f;
            let x0 = x * f;
            for sy in y0..(y0 + f).min(h) {
                for sx in x0..(x0 + f).min(w) {
                    let o = ((sy * w + sx) * 4) as usize;
                    r += src[o] as u64;
                    g += src[o + 1] as u64;
                    b += src[o + 2] as u64;
                    a += src[o + 3] as u64;
                    n += 1;
                }
            }
            let o = ((y * dw + x) * 4) as usize;
            out[o] = js_round(r as f64 / n as f64) as u8;
            out[o + 1] = js_round(g as f64 / n as f64) as u8;
            out[o + 2] = js_round(b as f64 / n as f64) as u8;
            out[o + 3] = js_round(a as f64 / n as f64) as u8;
        }
    }
    (out, dw, dh)
}

/// Bilinear-upsample a buffer into a destination sized w×h (deterministic) —
/// port of `upsampleBilinear`.
#[allow(clippy::too_many_arguments)]
pub fn upsample_bilinear(
    src: &[u8],
    sw: u32,
    sh: u32,
    dst: &mut [u8],
    w: u32,
    h: u32,
) {
    let sx_scale = sw as f64 / w as f64;
    let sy_scale = sh as f64 / h as f64;
    for y in 0..h {
        let sy = (((y as f64 + 0.5) * sy_scale - 0.5).max(0.0)).min((sh - 1) as f64);
        let y0 = sy.floor() as u32;
        let y1 = (y0 + 1).min(sh - 1);
        let fy = sy - y0 as f64;
        for x in 0..w {
            let sx = (((x as f64 + 0.5) * sx_scale - 0.5).max(0.0)).min((sw - 1) as f64);
            let x0 = sx.floor() as u32;
            let x1 = (x0 + 1).min(sw - 1);
            let fx = sx - x0 as f64;
            let o00 = ((y0 * sw + x0) * 4) as usize;
            let o10 = ((y0 * sw + x1) * 4) as usize;
            let o01 = ((y1 * sw + x0) * 4) as usize;
            let o11 = ((y1 * sw + x1) * 4) as usize;
            let do_ = ((y * w + x) * 4) as usize;
            for c in 0..4 {
                let top = src[o00 + c] as f64 + (src[o10 + c] as f64 - src[o00 + c] as f64) * fx;
                let bot = src[o01 + c] as f64 + (src[o11 + c] as f64 - src[o01 + c] as f64) * fx;
                dst[do_ + c] = js_round(top + (bot - top) * fy) as u8;
            }
        }
    }
}
